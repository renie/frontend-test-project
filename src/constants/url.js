export const getAPIUrl = endpoint => `https://jsonplaceholder.typicode.com/${endpoint}`;
