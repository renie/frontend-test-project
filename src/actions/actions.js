import { getAPIUrl } from '../constants/url.js'
import {
  GET_USERS_REQUEST,
  GET_USERS_SUCCESS,
  GET_USERS_ERROR
} from '../constants/actionTypes.js'


const requestUsers = () => ({ type: GET_USERS_REQUEST });
const receiveUsers = users => ({ type: GET_USERS_SUCCESS, users });
const errorUsers = error => ({ type: GET_USERS_ERROR, error });

export const fetchUsers = () => dispatch => {
  dispatch(requestUsers());

  return fetch(getAPIUrl('users'))
      .then(
        response => response.json(),
        error => dispatch(errorUsers(error))
      )
      .then(
        json => dispatch(receiveUsers(json))
      )
      .catch(
        error => dispatch(errorUsers(error)))
}
