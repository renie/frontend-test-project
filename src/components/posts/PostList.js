import React, { Component } from 'react';
import { connect } from 'react-redux';
import Post from './Post'
import { fetchPosts } from '../../actions/actions';
import './postList.css'

export class PostList extends Component {
  componentDidMount() {
    this.props.fetchPosts(this.props.userId)
  }

  render() {
    const user = this.props.users.find(user => user.id === this.props.userId)
    const posts = user ? user.posts : []

    return (
      <div className="postListContainer">
        <h3 className="label">Posts:</h3>
        <ul className="postList">
          {
            posts.map((post, index) => (
              <Post key={post.id} post={post} userId={user.id} />
            ))
          }
        </ul>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  users: state.appReducer.app.users
})

const mapDispatchToProps = dispatch => ({
  fetchPosts: userId => dispatch(fetchPosts(userId))
})

export default connect(mapStateToProps, mapDispatchToProps)(PostList)
