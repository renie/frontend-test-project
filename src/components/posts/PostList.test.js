import React from 'react'
import ReactDOM from 'react-dom'
import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { shallow } from 'enzyme'
import fetchMock from 'fetch-mock'

import { PostList } from './PostList'
import { fetchPosts } from '../../actions/actions'
import { getAPIUrl } from '../../constants/url.js'


const mockStore = configureMockStore([thunk])

describe('PostList', () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it('Create Component', () => {
    const expectedBody = [];
    const store = mockStore({ app: { users: [] }})
    const mockfetchPost = () => store.dispatch(fetchPosts(id))
    const id = 1

    fetchMock.getOnce(getAPIUrl('posts', id), {
      body: expectedBody,
      headers: { 'content-type': 'application/json' }
    })

    const wrapper = shallow(<PostList store={store} fetchPosts={mockfetchPost} userId={id} users={[]} />)

    expect(wrapper).toMatchSnapshot();
  })
})
