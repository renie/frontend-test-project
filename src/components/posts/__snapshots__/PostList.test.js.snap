// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostList Create Component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PostList
    fetchPosts={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    userId={1}
    users={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h3
          className="label"
        >
          Posts:
        </h3>,
        <ul
          className="postList"
        />,
      ],
      "className": "postListContainer",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Posts:",
          "className": "label",
        },
        "ref": null,
        "rendered": "Posts:",
        "type": "h3",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "postList",
        },
        "ref": null,
        "rendered": Array [],
        "type": "ul",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h3
            className="label"
          >
            Posts:
          </h3>,
          <ul
            className="postList"
          />,
        ],
        "className": "postListContainer",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Posts:",
            "className": "label",
          },
          "ref": null,
          "rendered": "Posts:",
          "type": "h3",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "postList",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ul",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
